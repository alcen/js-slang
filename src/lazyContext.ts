import { Context } from './types'

/**
 * Checks whether the interpreter should evaluate
 * the expression lazily in this context. Currently,
 * lazy evaluation is only implemented for Source
 * chapters 1 and 2, so lazyEvaluate will return true
 * only when the chapter is 1 or 2.
 * @param context The context to be checked.
 * @returns True, if the interpreter is to evaluate
 *     lazily. False, if the interpreter should
 *     evaluate eagerly.
 */
export default function lazyEvaluate(context: Context) {
  return lazyEvaluateInChapter(context.chapter)
}

/**
 * Checks whether the interpreter should evaluate the
 * expression lazily, given a chapter number.
 * @param chapter The chapter to be checked.
 * @returns True, if the interpreter is to evaluate
 *     lazily. False, if the interpreter should
 *     evaluate eagerly.
 */
export function lazyEvaluateInChapter(chapter: number) {
  return chapter === -1
}
