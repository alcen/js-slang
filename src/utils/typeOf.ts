import { Value } from "../types"

// We need to define our own typeof in order for null/array to display properly in error messages
export const typeOf = (v: Value) => {
  if (v === null) {
    return 'null'
  } else if (Array.isArray(v)) {
    return 'array'
  } else {
    return typeof v
  }
}
