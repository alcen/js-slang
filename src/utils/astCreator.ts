import * as es from 'estree'
import { astThunkNativeTag } from '../stdlib/lazy'
import { AllowedDeclarations, BlockExpression, FunctionDeclarationExpression } from '../types'
import { typeOf } from './typeOf'

export const locationDummyNode = (line: number, column: number) =>
  literal('Dummy', { start: { line, column }, end: { line, column } })

export const identifier = (name: string): es.Identifier => ({
  type: 'Identifier',
  name
})

/**
 * Constructs a new Thunk value for lazy evaluation. This function
 * does this by taking the original Literal in the Abstract Syntax
 * Tree, and changing it into an ArrowFunctionExpression with the
 * literal in its body
 * @param node The Literal to be transformed.
 */
export const mutateToThunk = (node: es.Literal): es.ObjectExpression => {
  // creates a new copy of node.loc
  function copyLoc() {
    if (!node.loc) {
      return undefined
    } else {
      return {
        start: Object.assign({}, node.loc.start),
        end: Object.assign({}, node.loc.end)
      }
    }
  }
  // make a new object for the new Thunk
  const newNode = node as any
  newNode.type = 'ObjectExpression'
  // get type of literal
  const type = typeOf(node.value)
  // ensure that the raw type becomes a string, not a variable
  const typeRaw = '"' + type + '"'
  // get rid of old value and old raw
  const oldValue = newNode.value
  newNode.value = undefined
  const oldRaw = newNode.raw
  newNode.raw = undefined
  newNode.properties = []
  // set Thunk properties
  // first, create type property
  newNode.properties[0] = {
    type: 'Property',
    start: newNode.start,
    end: newNode.end,
    loc: copyLoc(),
    method: false,
    shorthand: false,
    computed: false,
    key: {
      type: 'Identifier',
      start: newNode.start,
      end: newNode.end,
      loc: copyLoc(),
      name: 'type'
    },
    // value of type is a string literal
    value: {
      type: 'Literal',
      start: newNode.start,
      end: newNode.end,
      loc: copyLoc(),
      value: type,
      raw: typeRaw
    },
    kind: 'init'
  }
  // get arrow function
  const arrowFunction = {
    type: 'ArrowFunctionExpression',
    start: newNode.start,
    end: newNode.end,
    loc: copyLoc(),
    id: null,
    expression: true,
    generator: false,
    params: [],
    body: {
      type: 'Literal',
      start: newNode.start,
      end: newNode.end,
      loc: copyLoc(),
      value: oldValue,
      raw: oldRaw
    }
  }
  // then, create the value property
  // also create lambda function storing original literal
  newNode.properties[1] = {
    type: 'Property',
    start: newNode.start,
    end: newNode.end,
    loc: copyLoc(),
    method: false,
    shorthand: false,
    computed: false,
    key: {
      type: 'Identifier',
      start: newNode.start,
      end: newNode.end,
      loc: copyLoc(),
      name: 'value'
    },
    // value of 'value' is an ArrowFunctionExpression
    value: arrowFunction,
    kind: 'init'
  }
  // get toString to show the normal result
  const toString = {
    type: 'ArrowFunctionExpression',
    start: newNode.start,
    end: newNode.end,
    loc: copyLoc(),
    id: null,
    expression: true,
    generator: false,
    params: [],
    body: {
      type: 'Literal',
      start: newNode.start,
      end: newNode.end,
      loc: copyLoc(),
      value: JSON.stringify(oldValue),
      raw: JSON.stringify(oldRaw)
    },
    // add tag to prevent toString() from getting wrapped in
    // wrapArrowFunctionsToAllowNormalCallsAndNiceToString
    // (toString is already giving it a nice string representation)
    tag: astThunkNativeTag
  }
  // lastly create the toString property
  // so thunks appear as normal values
  newNode.properties[2] = {
    type: 'Property',
    start: newNode.start,
    end: newNode.end,
    loc: copyLoc(),
    method: false,
    shorthand: false,
    computed: false,
    key: {
      type: 'Identifier',
      start: newNode.start,
      end: newNode.end,
      loc: copyLoc(),
      name: 'toString'
    },
    value: toString,
    kind: 'init'
  }
  return newNode
}

export const literal = (value: string | number | boolean, loc?: es.SourceLocation): es.Literal => ({
  type: 'Literal',
  value,
  loc
})

export const memberExpression = (
  object: es.Expression,
  propertyString: string
): es.MemberExpression => ({
  type: 'MemberExpression',
  object,
  computed: false,
  property: identifier(propertyString)
})

export const declaration = (
  name: string,
  kind: AllowedDeclarations,
  init: es.Expression,
  loc?: es.SourceLocation
): es.VariableDeclaration => ({
  type: 'VariableDeclaration',
  declarations: [
    {
      type: 'VariableDeclarator',
      id: identifier(name),
      init
    }
  ],
  kind,
  loc
})

export const constantDeclaration = (name: string, init: es.Expression, loc?: es.SourceLocation) =>
  declaration(name, 'const', init, loc)

export const callExpression = (
  callee: es.Expression,
  args: es.Expression[],
  loc?: es.SourceLocation
): es.CallExpression => ({
  type: 'CallExpression',
  callee,
  arguments: args,
  loc
})

export const expressionStatement = (expression: es.Expression): es.ExpressionStatement => ({
  type: 'ExpressionStatement',
  expression
})

export const blockArrowFunction = (
  params: es.Identifier[],
  body: es.Statement[] | es.BlockStatement,
  loc?: es.SourceLocation
): es.ArrowFunctionExpression => ({
  type: 'ArrowFunctionExpression',
  expression: false,
  generator: false,
  params,
  body: Array.isArray(body) ? blockStatement(body) : body,
  loc
})

export const functionExpression = (
  params: es.Identifier[],
  body: es.Statement[] | es.BlockStatement,
  loc?: es.SourceLocation
): es.FunctionExpression => ({
  type: 'FunctionExpression',
  id: null,
  async: false,
  generator: false,
  params,
  body: Array.isArray(body) ? blockStatement(body) : body,
  loc
})

export const blockStatement = (body: es.Statement[]): es.BlockStatement => ({
  type: 'BlockStatement',
  body
})

export const program = (body: es.Statement[]): es.Program => ({
  type: 'Program',
  sourceType: 'module',
  body
})

export const returnStatement = (
  argument: es.Expression,
  loc?: es.SourceLocation
): es.ReturnStatement => ({
  type: 'ReturnStatement',
  argument,
  loc
})

export const property = (key: string, value: es.Expression): es.Property => ({
  type: 'Property',
  method: false,
  shorthand: false,
  computed: false,
  key: identifier(key),
  value,
  kind: 'init'
})

export const objectExpression = (properties: es.Property[]): es.ObjectExpression => ({
  type: 'ObjectExpression',
  properties
})

export const mutateToCallExpression = (
  node: es.Node,
  callee: es.Expression,
  args: es.Expression[]
) => {
  node.type = 'CallExpression'
  node = node as es.CallExpression
  node.callee = callee
  node.arguments = args
}

export const logicalExpression = (
  operator: es.LogicalOperator,
  left: es.Expression,
  right: es.Expression,
  loc?: es.SourceLocation
): es.LogicalExpression => ({
  type: 'LogicalExpression',
  operator,
  left,
  right,
  loc
})

export const conditionalExpression = (
  test: es.Expression,
  consequent: es.Expression,
  alternate: es.Expression,
  loc?: es.SourceLocation
): es.ConditionalExpression => ({
  type: 'ConditionalExpression',
  test,
  consequent,
  alternate,
  loc
})

export const arrayExpression = (elements: es.Expression[]): es.ArrayExpression => ({
  type: 'ArrayExpression',
  elements
})

export const assignmentExpression = (
  left: es.Identifier | es.MemberExpression,
  right: es.Expression
): es.AssignmentExpression => ({
  type: 'AssignmentExpression',
  operator: '=',
  left,
  right
})

export const binaryExpression = (
  operator: es.BinaryOperator,
  left: es.Expression,
  right: es.Expression,
  loc?: es.SourceLocation
): es.BinaryExpression => ({
  type: 'BinaryExpression',
  operator,
  left,
  right,
  loc
})

export const unaryExpression = (
  operator: es.UnaryOperator,
  argument: es.Expression,
  loc?: es.SourceLocation
): es.UnaryExpression => ({
  type: 'UnaryExpression',
  operator,
  prefix: true,
  argument,
  loc
})

// primitive: undefined is a possible value
export const primitive = (value: any): es.Expression => {
  return value === undefined ? identifier('undefined') : literal(value)
}

export const functionDeclarationExpression = (
  id: es.Identifier,
  params: es.Pattern[],
  body: es.BlockStatement,
  loc?: es.SourceLocation
): FunctionDeclarationExpression => ({
  type: 'FunctionExpression',
  id,
  params,
  body,
  loc
})

export const functionDeclaration = (
  id: es.Identifier | null,
  params: es.Pattern[],
  body: es.BlockStatement,
  loc?: es.SourceLocation
): es.FunctionDeclaration => ({
  type: 'FunctionDeclaration',
  id,
  params,
  body,
  loc
})

export const blockExpression = (
  body: es.Statement[],
  loc?: es.SourceLocation
): BlockExpression => ({
  type: 'BlockExpression',
  body,
  loc
})

export const arrowFunctionExpression = (
  params: es.Pattern[],
  body: es.Expression | es.BlockStatement,
  loc?: es.SourceLocation
): es.ArrowFunctionExpression => ({
  type: 'ArrowFunctionExpression',
  expression: body.type !== 'BlockStatement',
  generator: false,
  params,
  body,
  loc
})

export const variableDeclaration = (
  declarations: es.VariableDeclarator[],
  loc?: es.SourceLocation
): es.VariableDeclaration => ({
  type: 'VariableDeclaration',
  kind: 'const',
  declarations,
  loc
})

export const variableDeclarator = (
  id: es.Pattern,
  init: es.Expression,
  loc?: es.SourceLocation
): es.VariableDeclarator => ({
  type: 'VariableDeclarator',
  id,
  init,
  loc
})

export const ifStatement = (
  test: es.Expression,
  consequent: es.BlockStatement,
  alternate: es.Statement,
  loc?: es.SourceLocation
): es.IfStatement => ({
  type: 'IfStatement',
  test,
  consequent,
  alternate,
  loc
})
