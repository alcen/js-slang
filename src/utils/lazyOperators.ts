import { BinaryOperator, UnaryOperator, LogicalOperator } from 'estree'
import { JSSLANG_PROPERTIES } from '../constants'
import {
  CallingNonFunctionValue,
  ExceptionError,
  GetInheritedPropertyError,
  InvalidNumberOfArguments
} from '../errors/errors'
import { RuntimeSourceError } from '../errors/runtimeSourceError'
import {
  PotentialInfiniteLoopError,
  PotentialInfiniteRecursionError
} from '../errors/timeoutErrors'
import {
  makeThunk,
  makeThunkWithPrimitiveBinary,
  Thunk,
  makeThunkWithPrimitiveUnary,
  makeConditionalThunk
} from '../stdlib/lazy'
import { callExpression, locationDummyNode } from './astCreator'
import * as create from './astCreator'
import * as rttc from './rttc'

export function throwIfTimeout(start: number, current: number, line: number, column: number) {
  if (current - start > JSSLANG_PROPERTIES.maxExecTime) {
    throw new PotentialInfiniteLoopError(create.locationDummyNode(line, column))
  }
}

export function callIfFuncAndRightArgs(
  candidate: any,
  line: number,
  column: number,
  ...args: any[]
) {
  const dummy = create.callExpression(create.locationDummyNode(line, column), args, {
    start: { line, column },
    end: { line, column }
  })
  if (typeof candidate === 'function') {
    if (candidate.transformedFunction === undefined) {
      try {
        return candidate(...args)
      } catch (error) {
        // if we already handled the error, simply pass it on
        if (!(error instanceof RuntimeSourceError || error instanceof ExceptionError)) {
          throw new ExceptionError(error, dummy.loc!)
        } else {
          throw error
        }
      }
    } else {
      const expectedLength = candidate.transformedFunction.length
      const receivedLength = args.length
      if (expectedLength !== receivedLength) {
        throw new InvalidNumberOfArguments(dummy, expectedLength, receivedLength)
      }
      return candidate(...args)
    }
  } else {
    throw new CallingNonFunctionValue(candidate, dummy)
  }
}

export function boolOrErr(candidate: Thunk<any>, line: number, column: number) {
  const error = rttc.checkIfStatementT(create.locationDummyNode(line, column), candidate)
  if (error === undefined) {
    return candidate.toString()
  } else {
    return error
  }
}

export function unaryOp(
  operator: UnaryOperator,
  argument: Thunk<any>,
  line: number,
  column: number
) {
  const resultType = rttc.checkUnaryExpressionT(
    create.locationDummyNode(line, column),
    operator,
    argument
  )
  if (typeof resultType === 'string') {
    return evaluateUnaryExpression(operator, argument, resultType)
  } else {
    throw resultType
  }
}

/**
 * Delays evaluation of an unary expression
 * in Lazy Source, returning a Thunk.
 *
 * @param operator String representing the operator
 *     to be executed.
 * @param value The argument of this operator.
 * @param returnType The return type of the evaluated
 *     expression.
 */
export function evaluateUnaryExpression(
  operator: UnaryOperator,
  value: Thunk<any>,
  // default value '' to prevent problems with substitutor, intepreter
  returnType: string = ''
) {
  if (operator === '!') {
    return makeThunkWithPrimitiveUnary(value, x => !x, returnType, operator)
  } else if (operator === '-') {
    return makeThunkWithPrimitiveUnary(value, x => -x, returnType, operator)
  } else if (operator === '+') {
    return makeThunkWithPrimitiveUnary(value, x => +x, returnType, operator)
  } else {
    return makeThunk(undefined)
  }
}

export function binaryOp(
  operator: BinaryOperator,
  left: Thunk<any>,
  right: Thunk<any>,
  line: number,
  column: number
) {
  const resultType = rttc.checkBinaryExpressionT(
    create.locationDummyNode(line, column),
    operator,
    left,
    right
  )
  if (typeof resultType === 'string') {
    return evaluateBinaryExpression(operator, left, right, resultType)
  } else {
    throw resultType
  }
}

/**
 * Delays evaluation of a binary expression in
 * Lazy Source, returning a Thunk.
 *
 * @param operator String representing the operator
 *     to be executed.
 * @param left The first argument, or the left, of
 *     this operator.
 * @param right The second argument, or the right, of
 *     this operator.
 * @param returnType The return type of the evaluated
 *     expression.
 */
export function evaluateBinaryExpression(
  operator: BinaryOperator,
  left: Thunk<any>,
  right: Thunk<any>,
  // default value '' to prevent problems with substitutor, intepreter
  returnType: string = ''
): Thunk<any> {
  switch (operator) {
    case '+':
      return makeThunkWithPrimitiveBinary(left, right, (x, y) => x + y, returnType, operator)
    case '-':
      return makeThunkWithPrimitiveBinary(left, right, (x, y) => x - y, returnType, operator)
    case '*':
      return makeThunkWithPrimitiveBinary(left, right, (x, y) => x * y, returnType, operator)
    case '/':
      return makeThunkWithPrimitiveBinary(left, right, (x, y) => x / y, returnType, operator)
    case '%':
      return makeThunkWithPrimitiveBinary(left, right, (x, y) => x % y, returnType, operator)
    case '===':
      return makeThunkWithPrimitiveBinary(left, right, (x, y) => x === y, returnType, operator)
    case '!==':
      return makeThunkWithPrimitiveBinary(left, right, (x, y) => x !== y, returnType, operator)
    case '<=':
      return makeThunkWithPrimitiveBinary(left, right, (x, y) => x <= y, returnType, operator)
    case '<':
      return makeThunkWithPrimitiveBinary(left, right, (x, y) => x < y, returnType, operator)
    case '>':
      return makeThunkWithPrimitiveBinary(left, right, (x, y) => x > y, returnType, operator)
    case '>=':
      return makeThunkWithPrimitiveBinary(left, right, (x, y) => x >= y, returnType, operator)
    default:
      return makeThunk(undefined)
  }
}

/**
 * This function will be called in place of logical
 * operations like && (and) or || (or), in order to
 * check whether the Thunks on left and right are of
 * type boolean, and to execute it lazily
 * @param operator String representing the operator
 *     to be executed.
 * @param left The (boolean) expression on the left
 *     of the operator (Thunk)
 * @param right The (boolean) expression on the right
 *     of the operator (Thunk)
 * @param line Line number of the expression in
 *     the program
 * @param column Column number of the expression
 *     in the program
 */
export function logicalOp(
  operator: LogicalOperator,
  left: Thunk<any>,
  right: Thunk<any>,
  line: number,
  column: number
) {
  const leftType = boolOrErr(left, line, column)
  const rightType = boolOrErr(right, line, column)
  if (typeof leftType === 'string' && typeof rightType === 'string') {
    return evaluateLogicalExpression(operator, left, right)
  } else if (typeof leftType === 'string') {
    throw rightType
  } else {
    throw leftType
  }
}

/**
 * Delays evaluation of a logical expression && (and)
 * and || (or) in Lazy Source, returning a Thunk.
 *
 * @param operator String representing the operator
 *     to be executed.
 * @param left The first argument, or the left, of
 *     this operator.
 * @param right The second argument, or the right, of
 *     this operator.
 */
export function evaluateLogicalExpression(
  operator: LogicalOperator,
  left: Thunk<any>,
  right: Thunk<any>
): Thunk<any> {
  // string representation of resultant thunk
  const stringRep = left.toString() + ' ' + operator + ' ' + right.toString()
  switch (operator) {
    case '&&':
      return makeConditionalThunk(left, right, makeThunk(false), stringRep)
    case '||':
      return makeConditionalThunk(left, makeThunk(true), right, stringRep)
    default:
      return makeThunk(undefined)
  }
}

/**
 * This function will be called in place of conditional
 * expressions, in order to check whether the predicate
 * thunk is of type boolean, and to execute it lazily
 * @param predicate Predicate expression to be evaluated.
 * @param consequent Consequent to be evaluated, if
 *     predicate evaluates to true.
 * @param alternate Alternate to be evaluated, if
 *     predicate evaluates to false.
 * @param line Line number of the expression in
 *     the program
 * @param column Column number of the expression
 *     in the program
 */
export function conditionalOp(
  predicate: Thunk<any>,
  consequent: Thunk<any>,
  alternate: Thunk<any>,
  line: number,
  column: number
): Thunk<any> {
  const predicateType = boolOrErr(predicate, line, column)
  if (typeof predicateType === 'string') {
    return makeConditionalThunk(predicate, consequent, alternate)
  } else {
    throw predicateType
  }
}

/**
 * Limitations for current properTailCalls implementation:
 * Obviously, if objects ({}) are reintroduced,
 * we have to change this for a more stringent check,
 * as isTail and transformedFunctions are properties
 * and may be added by Source code.
 */
export const callIteratively = (f: any, ...args: any[]) => {
  let line = -1
  let column = -1
  const MAX_TIME = JSSLANG_PROPERTIES.maxExecTime
  const startTime = Date.now()
  const pastCalls: [string, any[]][] = []
  while (true) {
    const dummy = locationDummyNode(line, column)
    if (Date.now() - startTime > MAX_TIME) {
      throw new PotentialInfiniteRecursionError(dummy, pastCalls)
    } else if (typeof f !== 'function') {
      throw new CallingNonFunctionValue(f, dummy)
    }
    if (f.transformedFunction! !== undefined) {
      f = f.transformedFunction
      const expectedLength = f.length
      const receivedLength = args.length
      if (expectedLength !== receivedLength) {
        throw new InvalidNumberOfArguments(
          callExpression(locationDummyNode(line, column), args, {
            start: { line, column },
            end: { line, column }
          }),
          expectedLength,
          receivedLength
        )
      }
    }
    let res
    try {
      res = f(...args)
    } catch (error) {
      // if we already handled the error, simply pass it on
      if (!(error instanceof RuntimeSourceError || error instanceof ExceptionError)) {
        throw new ExceptionError(error, dummy.loc!)
      } else {
        throw error
      }
    }
    if (res === null || res === undefined) {
      return res
    } else if (res.isTail === true) {
      f = res.function
      args = res.arguments
      line = res.line
      column = res.column
      pastCalls.push([res.functionName, args])
    } else if (res.isTail === false) {
      return res.value
    } else {
      return res
    }
  }
}

export const wrap = (f: (...args: any[]) => any, stringified: string) => {
  const wrapped = (...args: any[]) => callIteratively(f, ...args)
  wrapped.transformedFunction = f
  wrapped[Symbol.toStringTag] = () => stringified
  wrapped.toString = () => stringified
  return wrapped
}

export const setProp = (obj: any, prop: any, value: any, line: number, column: number) => {
  const dummy = locationDummyNode(line, column)
  const error = rttc.checkMemberAccess(dummy, obj, prop)
  if (error === undefined) {
    return (obj[prop] = value)
  } else {
    throw error
  }
}

export const getProp = (obj: any, prop: any, line: number, column: number) => {
  const dummy = locationDummyNode(line, column)
  const error = rttc.checkMemberAccess(dummy, obj, prop)
  if (error === undefined) {
    if (obj[prop] !== undefined && !obj.hasOwnProperty(prop)) {
      throw new GetInheritedPropertyError(dummy, obj, prop)
    } else {
      return obj[prop]
    }
  } else {
    throw error
  }
}
