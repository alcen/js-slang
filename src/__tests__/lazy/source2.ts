import { stripIndent } from '../../utils/formatters'
import { expectResult, expectDisplayResult } from '../../utils/testing'
import { LAZY_SOURCE_2 } from '../../lazyContext'
import { ExecutionMethod } from '../../types'

// ================= UTILITY FUNCTIONS =================
const runTestSuccess = (
  executionMethod: ExecutionMethod,
  program: string,
  result: string,
  testName: string
) =>
  test(testName + ' (' + executionMethod + ')', () => {
    return expectResult(stripIndent(program), {
      chapter: LAZY_SOURCE_2,
      executionMethod
    }).toMatchInlineSnapshot(result)
  })

const runTestForDisplay = (
  executionMethod: ExecutionMethod,
  program: string,
  displayed: string,
  testName: string
) =>
  test(testName + ' (' + executionMethod + ')', () => {
    return expectDisplayResult(stripIndent(program), {
      chapter: LAZY_SOURCE_2,
      executionMethod
    }).toMatchInlineSnapshot(displayed)
  })

// ==================== LAZY TESTS =====================
const mapReturnsCorrectly = (executionMethod: ExecutionMethod) =>
  runTestSuccess(
    executionMethod,
    `
force_pair(map(x => x / 2, list(2, 4, 8)));
`,
    `
Array [
  1, Array [
    2, Array [
      4, null
    ]
  ]
]
  `,
    'map returns the correct list'
  )

const filterReturnsCorrectly = (executionMethod: ExecutionMethod) =>
  runTestSuccess(
    executionMethod,
    `
force_pair(filter(x => x % 2 === 0, list(1, 2, 3, 4, 5, 6)));
`,
    `
Array [
  2, Array [
    4, Array [
      6, null
    ]
  ]
]
  `,
    'filter returns the correct list'
  )

const listsAreMemoised = (executionMethod: ExecutionMethod) =>
  runTestForDisplay(
    executionMethod,
    `
  const ones = pair(1, ones);
  const twos = map(x => {
          display("mapped");
          return x * 2;
      }, ones);
  force(head(twos));
  force(head(twos));
  head(twos);
`,
    `
Array [
  "\\"mapped\\"",
]
  `,
    'lists are memoised'
  )

const listsAreMappedLazily = (executionMethod: ExecutionMethod) =>
  runTestForDisplay(
    executionMethod,
    `
  const ones = pair(1, ones);
  const twos = map(x => {
          display("mapped");
          return x * 2;
      }, ones);
  force(head(twos));
  head(tail(twos));
`,
    `
Array [
  "\\"mapped\\"",
  "\\"mapped\\""
]
  `,
    'lists are mapped lazily'
  )

// ================== RUN THE TESTS ====================
const testArray = [
  mapReturnsCorrectly,
  filterReturnsCorrectly,
  listsAreMemoised,
  listsAreMappedLazily
]
testArray.forEach(tst => tst('interpreter'))
testArray.forEach(tst => tst('native'))
testArray.forEach(tst => tst('auto'))
