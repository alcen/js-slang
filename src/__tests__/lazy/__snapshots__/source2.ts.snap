// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`filter returns the correct list (auto): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "equal(filter(x => x % 2 === 0, list(1, 2, 3, 4, 5, 6)), list(2, 4, 6));",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`filter returns the correct list (interpreter): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "equal(filter(x => x % 2 === 0, list(1, 2, 3, 4, 5, 6)), list(2, 4, 6));",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`filter returns the correct list (native): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "equal(filter(x => x % 2 === 0, list(1, 2, 3, 4, 5, 6)), list(2, 4, 6));",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`lists are mapped lazily (auto): expectDisplayResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "const ones = pair(1, ones);
const twos = map(x => {
        display(\\"mapped\\");
        return x * 2;
    }, ones);
force(head(twos));
head(tail(twos));",
  "displayResult": Array [
    "\\"mapped\\"",
    "\\"mapped\\"",
  ],
  "errors": Array [],
  "parsedErrors": "",
  "result": 2,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`lists are mapped lazily (interpreter): expectDisplayResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "const ones = pair(1, ones);
const twos = map(x => {
        display(\\"mapped\\");
        return x * 2;
    }, ones);
force(head(twos));
head(tail(twos));",
  "displayResult": Array [
    "\\"mapped\\"",
    "\\"mapped\\"",
  ],
  "errors": Array [],
  "parsedErrors": "",
  "result": 2,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`lists are mapped lazily (native): expectDisplayResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "const ones = pair(1, ones);
const twos = map(x => {
        display(\\"mapped\\");
        return x * 2;
    }, ones);
force(head(twos));
head(tail(twos));",
  "displayResult": Array [
    "\\"mapped\\"",
    "\\"mapped\\"",
  ],
  "errors": Array [],
  "parsedErrors": "",
  "result": 2,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`lists are memoised (auto): expectDisplayResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "const ones = pair(1, ones);
const twos = map(x => {
        display(\\"mapped\\");
        return x * 2;
    }, ones);
force(head(twos));
force(head(twos));
head(twos);",
  "displayResult": Array [
    "\\"mapped\\"",
  ],
  "errors": Array [],
  "parsedErrors": "",
  "result": 2,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`lists are memoised (interpreter): expectDisplayResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "const ones = pair(1, ones);
const twos = map(x => {
        display(\\"mapped\\");
        return x * 2;
    }, ones);
force(head(twos));
force(head(twos));
head(twos);",
  "displayResult": Array [
    "\\"mapped\\"",
  ],
  "errors": Array [],
  "parsedErrors": "",
  "result": 2,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`lists are memoised (native): expectDisplayResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "const ones = pair(1, ones);
const twos = map(x => {
        display(\\"mapped\\");
        return x * 2;
    }, ones);
force(head(twos));
force(head(twos));
head(twos);",
  "displayResult": Array [
    "\\"mapped\\"",
  ],
  "errors": Array [],
  "parsedErrors": "",
  "result": 2,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`map returns the correct list (auto): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "equal(map(x => x / 2, list(2, 4, 8)), list(1, 2, 4));",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`map returns the correct list (interpreter): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "equal(map(x => x / 2, list(2, 4, 8)), list(1, 2, 4));",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;

exports[`map returns the correct list (native): expectResult 1`] = `
Object {
  "alertResult": Array [],
  "code": "equal(map(x => x / 2, list(2, 4, 8)), list(1, 2, 4));",
  "displayResult": Array [],
  "errors": Array [],
  "parsedErrors": "",
  "result": true,
  "resultStatus": "finished",
  "visualiseListResult": Array [],
}
`;
